### PR TITLE
Update metric benchmarks

### PR DESCRIPTION
**Pull request by utpilla**
## Changes
- Update metrics benchmarks for API
- Use `iter_batched` method to generate the dynamic attributes before each iteration

### CODE REVIEW

File: opentelemetry/benches/metrics.rs [hunk 1]
```diff
@@ -51,16 +51,27 @@ fn counter_add(c: &mut Criterion) {
...
-        b.iter(|| {
-            let kv = vec![
-                KeyValue::new("attribute1", "value1"),
-                KeyValue::new("attribute2", "value2"),
-                KeyValue::new("attribute3", "value3"),
-                KeyValue::new("attribute4", "value4"),
-            ];
+        b.iter_batched(
+            || {
+                let value1 = black_box("a".repeat(6)); // Repeat character six times to match the length of value strings used in other benchmarks
```
Comment (lalitb) [inline opentelemetry/benches/metrics.rs:56]: nit - do we need black_box for creating these strings?

**Review by lalitb — APPROVED** [review]
Thanks for fixing these. Should we also add the benchmarks results as done here -  https://github.com/open-telemetry/opentelemetry-rust/blob/ad990d6d33aa73b35e0e3b792b1fe3ea4251b92b/opentelemetry/benches/anyvalue.rs#L6-L10

also add the machine configuration used for testing.

File: opentelemetry/benches/metrics.rs [hunk 2]
```diff
@@ -28,10 +28,10 @@ fn counter_add(c: &mut Criterion) {
             counter.add(
                 1,
                 &[
-                    KeyValue::new("attribute1", "value1"),
-                    KeyValue::new("attribute2", "value2"),
-                    KeyValue::new("attribute3", "value3"),
-                    KeyValue::new("attribute4", "value4"),
+                    black_box(KeyValue::new("attribute1", "value1")),
```
Comment (cijothomas) [inline opentelemetry/benches/metrics.rs:31]: curious : what is the actual need for this here? are we worried that compiler will do some optimizations here?

Comment (lalitb) [reply opentelemetry/benches/metrics.rs:31]: Good point. Should it be good to get the benchmark result with optimization (if any) as that would be the real scenario?

Comment (utpilla) [reply opentelemetry/benches/metrics.rs:31]: > are we worried that compiler will do some optimizations here?

Yes, that was my motivation to add `black_box`.

Comment (cijothomas) [reply opentelemetry/benches/metrics.rs:31]: i am not convinced we need to wrap in black_box.. If we were simply creating a KeyValue, then it makes sense, as compiler might think we are not using it anywhere so try to optimize in a way that does not match real scenarios.
In this case, Counter API is used, so we probably don't need black_box.

Comment (utpilla) [reply opentelemetry/benches/metrics.rs:31]: > compiler might think we are not using it anywhere so try to optimize in a way

The compiler could make other optimizations as well.

I benchmarked this method with and without `black_box` and there is a considerable perf difference around 32%

Without black_box: 13.523 ns
With black_box: 19.802

File: opentelemetry/benches/metrics.rs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-use criterion::{criterion_group, criterion_main, Criterion};
+use criterion::{black_box, criterion_group, criterion_main, BatchSize, Criterion};
```
Comment (cijothomas) [inline opentelemetry/benches/metrics.rs:1]: could you add the numbers here, similar to https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/benches/metric_counter.rs#L1-L13 for  easy reference.

File: opentelemetry/benches/metrics.rs (hunk 2 above)
Comment (cijothomas) [reply opentelemetry/benches/metrics.rs:31]: we want to measure the real user scenario right? Won't compiler be doing same optimization (the one wihtthout blackbox) in their code?

Comment (utpilla) [reply opentelemetry/benches/metrics.rs:31]: Yeah, it makes sense to mimic the real user scenario.

File: opentelemetry/benches/metrics.rs (hunk 1 above)
Comment (utpilla) [reply opentelemetry/benches/metrics.rs:56]: No, we can remove these as well.

File: opentelemetry/benches/metrics.rs (hunk 2 above)
Comment (utpilla) [reply opentelemetry/benches/metrics.rs]: Updated.

File: opentelemetry/benches/metrics.rs (hunk 3 above)
Comment (utpilla) [reply opentelemetry/benches/metrics.rs]: Updated.

File: opentelemetry/benches/metrics.rs [hunk 4]
```diff
@@ -44,23 +58,34 @@ fn counter_add(c: &mut Criterion) {
...
-            let kv = vec![
-                KeyValue::new("attribute1", "value1"),
-                KeyValue::new("attribute2", "value2"),
-                KeyValue::new("attribute3", "value3"),
-                KeyValue::new("attribute4", "value4"),
-            ];
+    c.bench_function("AddWithDynamicAttributes", |b| {
+        b.iter_batched(
+            || {
+                let value1 = "value1".to_string(); // Repeat character six times to match the length of value strings used in other benchmarks
```
Comment (cijothomas) [inline opentelemetry/benches/metrics.rs:70]:
```suggestion
                let value1 = "value1".to_string();
```

File: opentelemetry/benches/metrics.rs [hunk 5]
```diff
@@ -1,4 +1,19 @@
...
+    criterion = "0.5.1"
+    OS: Ubuntu 22.04.4 LTS (5.15.153.1-microsoft-standard-WSL2)
+    Hardware: Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz, 16vCPUs,
+    RAM: 64.0 GB
+    | Test                                                | Average time|
+    |-----------------------------------------------------|-------------|
+    | NoAttributes                                        | 1.1616 ns   |
+    | AddWithInlineStaticAttributes                       | 13.296 ns   |
+    | AddWithStaticArray                                  | 1.1612 ns   |
+    | AddWithDynamicAttributes                            | 110.40 ns   |
```
Comment (utpilla) [inline opentelemetry/benches/metrics.rs:12]: This benchmark method runs much faster when run in isolation. It takes around 80 ns instead of 110 ns.

**Pull request merged**